design-documents/atomics-ABI.rst

```diff
@@ -0,0 +1,308 @@
+
+Rationale Document for C11 Atomics ABI.
+***************************************
+
+Preamble
+========
+
+Background
+----------
+
+This document describes the rationale behind the ABI choices made for mapping
+from C11 atomic operations to Arm AArch64 assembly sequences.
+
+From the perspective of the Arm ABI we have some decisions to
+make:
+
+- We need to choose a baseline ABI (a set of mappings), that is compatible for all versions of the Armv8 architecture.
+- The mappings should cover atomic accesses of various sign, size, and type accessible through C11 atomic operations using compiler profiles.
+
+The main trade-offs we have identified or have been made aware of are:
+
+- Performance of different mappings versus compatibility with all architectures.
+- Whether certain compiler operations lead to unexpected behaviours.
+
+As motivated by the use cases expanded upon below:
+
+- The need for a baseline ABI
+- Knowing when an implementation departs from that baseline
+- Backwards compatibility of atomics as new mappings are added
+- Compatibility between compilers and runtimes
+- The need to constrain optimisations on specific atomic operations
+- Documenting the interoperable mappings
+- providing a basis upon which ABI compatibility can be tested.
+
+References
+----------
+
+This document refers to, or is referred to by, the following documents.
+
+.. table::
+
+  +-------------+--------------------------------------------------------------+-----------------------------------------------------------------------------+
+  | Ref         | External reference or URL                                    | Title                                                                       |
+  +=============+==============================================================+=============================================================================+
+  | ARMARM_     | DDI 0487                                                     | Arm Architecture Reference Manual Armv8 for Armv8-A architecture profile    |
+  +-------------+--------------------------------------------------------------+-----------------------------------------------------------------------------+
+  | PAPER_      | CGO paper                                                    | Compiler Testing with Relaxed Memory Models                                 |
+  +-------------+--------------------------------------------------------------+-----------------------------------------------------------------------------+
+
+
+
+Note: At the time of writing C23 is not released, as such ISO C17 is considered
+the latest published document.
+
+Use-cases known of so far
+-------------------------
+
+
+A Baseline: Describing current implementations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ABI we provide is a baseline specification that compilers should or do implement.
+The ABI provides a grounds to be compatible across all versions of the Armv8 architecture. Most
```

Comment:
"provides a grounds to be compatible across"  is a bit loose. I think you could use something like
Compilers that implement the baseline specification are compatible across all versions of the Armv8 architecture.